src/libos/crates/async-rt/src/sync/mutex/inner.rs

```diff
@@ -0,0 +1,199 @@
+///
+/// There are three states of this mutex:
+/// - Free: No one is holding the lock. Init state and can also be set by the thread who release the lock.
+/// - Locked: One thread is holding the lock. Set by the thread who acquires the lock.
+/// - LockedWithWaiters: One thread is holding the lock and some threads are waiting. Set by the waiting thread.
+#[derive(Debug)]
+pub(super) struct MutexInner {
+    status: AtomicLockStatus,
+    waiters: AtomicU64,
+    waiter_queue: WaiterQueue,
+}
+
+// This struct is the atomic wrapper for LockStatus.
+#[derive(Debug)]
+struct AtomicLockStatus(AtomicU64);
+
+#[derive(Debug, Copy, Clone)]
+#[repr(u64)]
+enum LockStatus {
+    Free = 0,
+    Locked = 1,
+    LockedWithWaiters = 2,
+}
+
+impl MutexInner {
+    pub(super) fn new() -> MutexInner {
+        MutexInner {
+            status: AtomicLockStatus::new(),
+            waiters: AtomicU64::new(0),
+            waiter_queue: WaiterQueue::new(),
+        }
+    }
+
+    pub(super) fn try_lock(&self) -> Result<()> {
+        if self.status.try_set_lock().is_ok() {
+            Ok(())
+        } else {
+            Err(errno!(EBUSY, "the lock is held by other threads"))
+        }
+    }
+
+    pub(super) async fn lock(&self) {
+        if let Ok(_) = self.try_lock() {
+            return;
+        }
+
+        // In musl's implmenetation, this is `100`. Considering more overhead in SGX environment,
+        // here we make it bigger.
+        const SPIN_COUNT: usize = 1000;
+
+        // Spin for a short while if no one is waiting but the lock is held.
+        let mut spins = SPIN_COUNT;
+        while spins != 0
+            && self.status.is_locked()
+            // Can't reorder here. `Relaxed` is enough.
+            && self.waiters.load(Ordering::Relaxed) == 0
+        {
+            hint::spin_loop();
+            spins -= 1;
+        }
+
+        loop {
+            if let Ok(_) = self.try_lock() {
+                return;
+            }
+
+            // In try_set_lock_with_waiters, `AcqRel` will make sure this increment happens before. Thus, `Relaxed` can be used here.
+            self.waiters.fetch_add(1, Ordering::Relaxed);
+
+            // Ignore the result here. If the state transition fails, the next wait will not block.
+            let _ = self.status.try_set_lock_with_waiters();
+
+            // Wait for unlock.
+            self.wait().await;
+
+            self.waiters.fetch_sub(1, Ordering::Relaxed);
+        }
+    }
+
+    pub(super) fn unlock(&self) {
+        // `get_current_status_and_set_free` will make sure this happens before. Thus, `Relaxed` can be used here.
+        let waiters = self.waiters.load(Ordering::Relaxed);
+        let previous_status = self.status.get_current_status_and_set_free();
+
+        if waiters != 0 {
+            self.wake_one_waiter();
+            return;
+        }
+
+        if previous_status.is_locked_with_waiters() {
+            panic!("the waiter number overflows to 0");
+        }
```

Comment:
This is incorrect. It is possible that `waiters == 0 && previous_status.is_locked_with_waiters()` without overflow. You can figure it out yourself... in no time.

Reply:
Got it. Remove this check. Done.